File: benchmarks/float_compute.py
import time

RUNS = 3

def sum_inv_squares(n):
    s = 0.0
    for i in range(1, n + 1):
        fi = float(i)
        s += 1.0 / (fi * fi)
    return s

def leibniz_pi(n):
    s = 0.0
    sign = 1.0
    for i in range(1, n + 1):
        denom = float(i * 2 - 1)
        s += sign / denom
        sign = -sign
    return s * 4.0

def mandelbrot_iter(cr, ci, max_iter):
    zr = 0.0
    zi = 0.0
    for i in range(max_iter):
        tr = zr * zr - zi * zi + cr
        zi = 2.0 * zr * zi + ci
        zr = tr
        if zr * zr + zi * zi > 4.0:
            return i
    return max_iter

def main():
    print("=== Python Float Benchmarks ===")

    # Benchmark 1: Basel problem
    n = 10000000
    sum_inv_squares(n)  # warmup
    best = float('inf')
    result = 0.0
    for run in range(RUNS):
        start = time.time_ns()
        r = sum_inv_squares(n)
        elapsed = time.time_ns() - start
        if elapsed < best:
            best = elapsed
            result = r
    print(f"Basel sum(1/i^2, 1..{n}):")
    print(f"  Time: {best // 1000000} ms")
    print(f"  Result: {result:.10f}")

    # Benchmark 2: Leibniz pi
    n = 10000000
    leibniz_pi(n)  # warmup
    best = float('inf')
    for run in range(RUNS):
        start = time.time_ns()
        r = leibniz_pi(n)
        elapsed = time.time_ns() - start
        if elapsed < best:
            best = elapsed
            result = r
    print(f"Leibniz pi({n} terms):")
    print(f"  Time: {best // 1000000} ms")
    print(f"  Result: {result:.10f}")

    # Benchmark 3: Mandelbrot grid
    size = 200
    # warmup
    t = 0
    for y in range(size):
        for x in range(size):
            cr = (x - 100.0) / 50.0
            ci = (y - 100.0) / 50.0
            t += mandelbrot_iter(cr, ci, 100)
    best = float('inf')
    total = 0
    for run in range(RUNS):
        start = time.time_ns()
        t = 0
        for y in range(size):
            for x in range(size):
                cr = (x - 100.0) / 50.0
                ci = (y - 100.0) / 50.0
                t += mandelbrot_iter(cr, ci, 100)
        elapsed = time.time_ns() - start
        if elapsed < best:
            best = elapsed
            total = t
    print(f"Mandelbrot {size}x{size}:")
    print(f"  Time: {best // 1000000} ms")
    print(f"  Result: {total}")

if __name__ == "__main__":
    main()

File: benchmarks/float_compute.js
const RUNS = 3;

function sumInvSquares(n) {
    let sum = 0.0;
    for (let i = 1; i <= n; i++) {
        sum += 1.0 / (i * i);
    }
    return sum;
}

function leibnizPi(n) {
    let sum = 0.0;
    let sign = 1.0;
    for (let i = 1; i <= n; i++) {
        const denom = i * 2 - 1;
        sum += sign / denom;
        sign = -sign;
    }
    return sum * 4.0;
}

function mandelbrotIter(cr, ci, maxIter) {
    let zr = 0.0, zi = 0.0;
    for (let i = 0; i < maxIter; i++) {
        const tr = zr * zr - zi * zi + cr;
        zi = 2.0 * zr * zi + ci;
        zr = tr;
        if (zr * zr + zi * zi > 4.0) {
            return i;
        }
    }
    return maxIter;
}

function distance(x1, y1, x2, y2) {
    const dx = x2 - x1;
    const dy = y2 - y1;
    return Math.sqrt(dx * dx + dy * dy);
}

function trigSum(n) {
    let sum = 0.0;
    for (let i = 0; i < n; i++) {
        const x = i * 0.001;
        sum += Math.sin(x) + Math.cos(x);
    }
    return sum;
}

function main() {
    console.log("=== Node.js Float Benchmarks ===");

    // Benchmark 1: Basel problem
    const n1 = 10000000;
    sumInvSquares(n1); // warmup
    let best = BigInt(Number.MAX_SAFE_INTEGER);
    let resultF;
    for (let run = 0; run < RUNS; run++) {
        let start = process.hrtime.bigint();
        let r = sumInvSquares(n1);
        let elapsed = process.hrtime.bigint() - start;
        if (elapsed < best) { best = elapsed; resultF = r; }
    }
    console.log(`Basel sum(1/i^2, 1..${n1}):`);
    console.log(`  Time: ${best / 1000000n} ms`);
    console.log(`  Result: ${resultF.toFixed(10)}`);

    // Benchmark 2: Leibniz pi
    const n2 = 10000000;
    leibnizPi(n2); // warmup
    best = BigInt(Number.MAX_SAFE_INTEGER);
    for (let run = 0; run < RUNS; run++) {
        let start = process.hrtime.bigint();
        let r = leibnizPi(n2);
        let elapsed = process.hrtime.bigint() - start;
        if (elapsed < best) { best = elapsed; resultF = r; }
    }
    console.log(`Leibniz pi(${n2} terms):`);
    console.log(`  Time: ${best / 1000000n} ms`);
    console.log(`  Result: ${resultF.toFixed(10)}`);

    // Benchmark 3: Mandelbrot grid
    const size = 200;
    // warmup
    {
        let t = 0;
        for (let y = 0; y < size; y++)
            for (let x = 0; x < size; x++)
                t += mandelbrotIter((x - 100.0) / 50.0, (y - 100.0) / 50.0, 100);
    }
    best = BigInt(Number.MAX_SAFE_INTEGER);
    let total;
    for (let run = 0; run < RUNS; run++) {
        let start = process.hrtime.bigint();
        let t = 0;
        for (let y = 0; y < size; y++) {
            for (let x = 0; x < size; x++) {
                const cr = (x - 100.0) / 50.0;
                const ci = (y - 100.0) / 50.0;
                t += mandelbrotIter(cr, ci, 100);
            }
        }
        let elapsed = process.hrtime.bigint() - start;
        if (elapsed < best) { best = elapsed; total = t; }
    }
    console.log(`Mandelbrot ${size}x${size}:`);
    console.log(`  Time: ${best / 1000000n} ms`);
    console.log(`  Result: ${total}`);

    // Benchmark 4: Distance computation
    const n4 = 10000000;
    // warmup
    {
        let s = 0.0;
        for (let i = 0; i < n4; i++) {
            const x = i * 0.001;
            s += distance(x, 0.0, 0.0, x);
        }
    }
    best = BigInt(Number.MAX_SAFE_INTEGER);
    for (let run = 0; run < RUNS; run++) {
        let start = process.hrtime.bigint();
        let s = 0.0;
        for (let i = 0; i < n4; i++) {
            const x = i * 0.001;
            s += distance(x, 0.0, 0.0, x);
        }
        let elapsed = process.hrtime.bigint() - start;
        if (elapsed < best) { best = elapsed; resultF = s; }
    }
    console.log(`Distance (${n4} iterations):`);
    console.log(`  Time: ${best / 1000000n} ms`);
    console.log(`  Result: ${resultF.toFixed(10)}`);

    // Benchmark 5: Trig computation
    const n5 = 5000000;
    trigSum(n5); // warmup
    best = BigInt(Number.MAX_SAFE_INTEGER);
    for (let run = 0; run < RUNS; run++) {
        let start = process.hrtime.bigint();
        let r = trigSum(n5);
        let elapsed = process.hrtime.bigint() - start;
        if (elapsed < best) { best = elapsed; resultF = r; }
    }
    console.log(`Trig sum(${n5}):`);
    console.log(`  Time: ${best / 1000000n} ms`);
    console.log(`  Result: ${resultF.toFixed(10)}`);
}

main();
